src/mscorlib/src/System/TimeZoneInfo.cs

```diff
@@ -1944,10 +1995,168 @@ private class OffsetAndRule {
+            byte[] buffer = new byte[rawData.Length];
+
+            foreach (string filePath in GetAllTimeZoneFilePaths(timeZoneDirectory))
+            {
+                // skip the localtime and posixrules file, since they won't give us the correct id
+                if (!string.Equals(filePath, localtimeFilePath, StringComparison.OrdinalIgnoreCase)
+                    && !string.Equals(filePath, posixrulesFilePath, StringComparison.OrdinalIgnoreCase))
+                {
+                    try
+                    {
+                        using (FileStream stream = new FileStream(filePath, FileMode.Open, FileAccess.Read))
+                        {
+                            if (stream.Length == rawData.Length)
+                            {
+                                bool isMatch = true;
+                                int index = 0;
+                                int count = rawData.Length;
+
+                                while (count > 0 && isMatch)
+                                {
+                                    int n = stream.Read(buffer, index, count);
+                                    if (n == 0)
+                                        __Error.EndOfFile();
+
+                                    int end = index + n;
+                                    for (int i = index; i < end; i++)
+                                    {
+                                        if (buffer[i] != rawData[i])
+                                        {
+                                            isMatch = false;
+                                            break;
+                                        }
+                                    }
+
+                                    index += n;
+                                    count -= n;
+                                }
+
+                                if (isMatch)
+                                {
+                                    // if all bytes are the same, this must be the right tz file
+                                    id = filePath;
+
+                                    // strip off the root time zone directory
+                                    if (id.StartsWith(timeZoneDirectory))
+                                    {
+                                        id = id.Substring(timeZoneDirectory.Length);
+                                    }
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                    // skip the file if it can't be opened
+                    catch (IOException) { }
+                    catch (SecurityException) { }
+                    catch (UnauthorizedAccessException) { }
+                }
+            }
+
+            return id;
+        }
+
+        private static IEnumerable<string> GetAllTimeZoneFilePaths(string timeZoneDirectory)
```

Comment:
It's possible for exceptions to be thrown during iteration (e.g. during calls to `enumerator.MoveNext()` as part of the `foreach` in the calling method). Which means this method should probably be removed and the try/catch moved up to the calling method, around the `foreach`.